lib/Service/RetentionService.php

```diff
@@ -0,0 +1,358 @@
+		if ($discoveryTimestamp === 0) {
+			// Set "now" as created at timestamp for the user.
+			$this->config->setUserValue($user->getUID(), 'user_retention', 'user_created_at', (string) $this->time->getTime());
+
+			throw new SkipUserException(
+				'New user, saving discovery time: {user}',
+				['user' => $user->getUID()]
+			);
+		}
+
+		return $discoveryTimestamp;
+	}
+
+	/**
+	 * @param IUser $user
+	 * @throws SkipUserException When the user is part of a group and should therefor be skipped
+	 */
+	protected function skipUserBasedOnProtectedGroupMembership(IUser $user): void {
+		if (empty($this->excludedGroups)) {
+			return;
+		}
+
+		$userGroups = $this->groupManager->getUserGroupIds($user);
+		$excludedGroups = array_intersect($userGroups, $this->excludedGroups);
+		if (!empty($excludedGroups)) {
+			throw new SkipUserException(
+				'Skipping user because of excluded groups ({groups}): {user}',
+				[
+					'user' => $user->getUID(),
+					'groups' => implode(',', $excludedGroups),
+				]
+			);
+		}
+	}
+
+	protected function getAuthTokensLastActivity(IUser $user): ?int {
+		/** @var Manager $authTokenManager */
+		$authTokenManager = $this->server->get(Manager::class);
+		/** @var PublicKeyToken[] $tokens */
+		$tokens = $authTokenManager->getTokenByUser($user->getUID());
+
+		$lastActivities = [];
+		foreach ($tokens as $token) {
+			$lastActivities[] = $token->getLastActivity();
+		}
+
+		if (empty($lastActivities)) {
+			return null;
+		}
+
+		return max(...$lastActivities);
+	}
+
+	protected function prepareLDAPUser(IUser $user): bool {
+		try {
+			$ldapProvider = $this->server->get(ILDAPProvider::class);
+			if($ldapProvider instanceof IDeletionFlagSupport) {
+				$ldapProvider->flagRecord($user->getUID());
+				$this->logger->info('Marking LDAP user as deleted: ' . $user->getUID());
+			}
+		} catch (\Exception $e) {
+			$this->logger->warning($e->getMessage(), [
+				'exception' => $e,
+			]);
+			return false;
+		}
+		return true;
+	}
+
+	protected function sendReminder(IUser $user, int $lastActivity, int $policyDays): void {
+		if (!$user->getEMailAddress()) {
+			$this->logger->warning('Could not send account retention reminder to {user} because no email address is configured.', [
+				'user' => $user->getUID(),
+			]);
+			return;
+		}
+
+		$this->logger->debug('Send reminder to account: {user}', [
+			'user' => $user->getUID(),
+		]);
+
+		$l = $this->l10nFactory->get('user_retention', $this->l10nFactory->getUserLanguage($user));
+
+		$message = $this->mailer->createMessage();
+		$template = $this->mailer->createEMailTemplate('user_retention.Reminder');
+		$template->setSubject($l->t('Important information regarding your account'));
+
+		$template->addHeader();
+		$template->addHeading($l->t('Account deletion'));
+		$template->addBodyText(str_replace('{date}', $l->l('date', $lastActivity), $l->t('You have used your account since {date}.')));
```

Comment:
```suggestion
		$template->addBodyText(str_replace('{date}', $l->l('date', $lastActivity), $l->t('You have not used your account since {date}.')));
```
No?